### PR TITLE
batch_prefill python api

### PR DESCRIPTION
**Pull request by hnyls2002**
A python api PR,

### CODE REVIEW

**Comment by hnyls2002** [comment]
Something still needs to be fixed and has yet to be tested. 

When the layout of kv_data in the `paged_kv_t` is HND (`[..., page_size, num_kv_heads, head_dim]`), is layout of q still `[sum(extend_lens), num_qo_heads, head_dim]` or `[num_qo_heads, sum(extend_lens), head_dim]` ? I am not sure about this.

File: python/csrc/batch_prefill.cu [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        static_cast<int32_t*>(kv_last_page_len.data_ptr()));
+    cudaError_t status =
+        BatchPrefillWithPagedKVCache<PageStorage::kIndices, c_type, c_type, int32_t>(
+            static_cast<c_type*>(q.data_ptr()), paged_kv,
+            static_cast<int32_t*>(q_indptr.data_ptr()), static_cast<c_type*>(o.data_ptr()),
+            /*tmp=*/nullptr,
+            /*lse=*/nullptr, num_qo_heads, causal, RotaryMode(rotary_mode),
+            allow_fp16_qk_reduction);
+    TORCH_CHECK(status == cudaSuccess, "BatchPrefillWithPagedKVCache failed with error code ",
+                status);
```
Comment (yzh119) [inline python/csrc/batch_prefill.cu]:
```suggestion
                cudaGetErrorString(status));
```

File: python/tests/test_shared_prefix_kernels.py [hunk 2]
```diff
@@ -32,5 +32,74 @@ def test_batch_decode_with_shared_prefix_padded_kv_cache(
...
+    q_indptr = torch.arange(0, batch_size + 1).to(0) * qo_len
+    num_pages_per_seq = (kv_len + page_size - 1) // page_size
+    total_num_pages = num_pages_per_seq * batch_size
+    kv_data = (
+        torch.randn(total_num_pages, 2, num_kv_heads, page_size, head_dim).to(0).half()
+    )
+    kv_indptr = torch.arange(0, batch_size + 1).to(0) * num_pages_per_seq
+    kv_indices = torch.arange(0, total_num_pages).to(0)
+    kv_last_page_len = torch.full(
+        (batch_size,), kv_len % page_size, dtype=torch.int32
```
Comment (yzh119) [inline python/tests/test_shared_prefix_kernels.py]: The` last_page_len`'s range should be [1, page_size]:
```suggestion
        (batch_size,), (kv_len - 1) % page_size + 1, dtype=torch.int32
```

**Review by yzh119 — COMMENTED** [review]
@hnyls2002 Thanks for the PR!
Please revise the PR according to the suggestions.

File: python/tests/test_shared_prefix_kernels.py [hunk 3]
```diff
@@ -32,5 +33,77 @@ def test_batch_decode_with_shared_prefix_padded_kv_cache(
...
+        o_ref_i = flashinfer.ops.single_prefill_with_kv_cache(qi, ki, vi, True)
+        o_i_np = o[q_indptr[i] : q_indptr[i + 1]].cpu().numpy()
+        o_ref_i_np = o_ref_i.cpu().numpy()
+        numpy.testing.assert_allclose(o_i_np, o_ref_i_np, rtol=1e-3, atol=1e-3)
+
+
 if __name__ == "__main__":
+    torch.manual_seed(0)
     test_batch_decode_with_shared_prefix_padded_kv_cache(12, 37, 54, 8, 128)
+    test_batch_prefill_with_paged_kv_cache(12, 54, 37, 8, 8, 8, 128)
```
Comment (yzh119) [inline python/tests/test_shared_prefix_kernels.py]: These tests are not related to `shared_prefix`, I suggest moving them to another file `test_batch_prefill_kernels.py`.

File: python/tests/test_shared_prefix_kernels.py [hunk 4]
```diff
@@ -32,5 +33,77 @@ def test_batch_decode_with_shared_prefix_padded_kv_cache(
...
+            dim=0,
+        )
+        o_ref_i = flashinfer.ops.single_prefill_with_kv_cache(qi, ki, vi, True)
+        o_i_np = o[q_indptr[i] : q_indptr[i + 1]].cpu().numpy()
+        o_ref_i_np = o_ref_i.cpu().numpy()
+        numpy.testing.assert_allclose(o_i_np, o_ref_i_np, rtol=1e-3, atol=1e-3)
+
+
 if __name__ == "__main__":
+    torch.manual_seed(0)
```
Comment (yzh119) [inline python/tests/test_shared_prefix_kernels.py]: This statement would not be triggered if we run pytest (which do not enter `if __name__ == "__main__":` condition), I suggest moving it to each test functions.

**Review by yzh119 — APPROVED** [review]
Overall LGTM, some minor issues.
